src/main/java/io/supertokens/inmemorydb/queries/TOTPQueries.java

```diff
@@ -0,0 +1,197 @@
+package io.supertokens.inmemorydb.queries;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+import io.supertokens.inmemorydb.Start;
+import io.supertokens.inmemorydb.config.Config;
+import io.supertokens.pluginInterface.RowMapper;
+import io.supertokens.pluginInterface.exceptions.StorageQueryException;
+import io.supertokens.pluginInterface.totp.TOTPDevice;
+import io.supertokens.pluginInterface.totp.TOTPUsedCode;
+
+import static io.supertokens.inmemorydb.QueryExecutorTemplate.execute;
+import static io.supertokens.inmemorydb.QueryExecutorTemplate.update;
+
+public class TOTPQueries {
+    public static String getQueryToCreateUserDevicesTable(Start start) {
+        return "CREATE TABLE IF NOT EXISTS " + Config.getConfig(start).getTotpUserDevicesTable() + " ("
+                + "user_id VARCHAR(128) NOT NULL," + "device_name VARCHAR(256) NOT NULL,"
+                + "secret_key VARCHAR(256) NOT NULL,"
+                + "period INTEGER NOT NULL," + "skew INTEGER NOT NULL," + "verified BOOLEAN NOT NULL,"
+                + "PRIMARY KEY (user_id, device_name))";
+    }
+
+    public static String getQueryToCreateUsedCodesTable(Start start) {
+        return "CREATE TABLE IF NOT EXISTS " + Config.getConfig(start).getTotpUsedCodesTable() + " ("
+                + "user_id VARCHAR(128) NOT NULL, " + "device_name VARCHAR(256), "
+                + "code CHAR(6) NOT NULL," + "is_valid_code BOOLEAN NOT NULL,"
+                + "expiry_time_ms BIGINT UNSIGNED NOT NULL," // Note: UNSIGNED won't work in Postgres
+                + "FOREIGN KEY (user_id, device_name) REFERENCES " + Config.getConfig(start).getTotpUserDevicesTable()
+                + "(user_id, device_name) ON DELETE CASCADE);";
+    }
+
+    public static String getQueryToCreateUsedCodesIndex(Start start) {
+        return "CREATE INDEX IF NOT EXISTS totp_used_codes_expiry_time_ms_index ON "
+                + Config.getConfig(start).getTotpUsedCodesTable() + " (expiry_time_ms)";
+    }
+
+    public static void createDevice(Start start, TOTPDevice device)
+            throws StorageQueryException, SQLException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " (user_id, device_name, secret_key, period, skew, verified) VALUES (?, ?, ?, ?, ?, ?)";
+
+        update(start, QUERY, pst -> {
+            pst.setString(1, device.userId);
+            pst.setString(2, device.deviceName);
+            pst.setString(3, device.secretKey);
+            pst.setInt(4, device.period);
+            pst.setInt(5, device.skew);
+            pst.setBoolean(6, device.verified);
+        });
+    }
+
```

Comment:
add unit test for if we are inserting into used codes with device_name as null - and make sure that it doesn't have foreign key violation.

Reply:
TODO